hyperon_das/traverse_engines.py

```diff
@@ -0,0 +1,197 @@
+                    continue
+                if target_type and target_type not in link['targets_type']:
+                    continue
+                if custom_filter and callable(custom_filter):
+                    ret = custom_filter(link)
+                    if not isinstance(ret, bool):
+                        raise TypeError('The function must return a boolean')
+                    if ret is False:
+                        continue
+                if handles_only:
+                    link = link['handle']
+                filtered_links.append(link)
+            links = filtered_links
+        return links
+
+    @abstractmethod
+    def get(self) -> Union[str, Dict[str, Any]]:
+        ...
+
+    @abstractmethod
+    def get_links(self, **kwargs) -> ListIterator:
+        ...
+
+    @abstractmethod
+    def get_neighbors(self, **kwargs) -> Union[List[str], List[Dict[str, Any]]]:
+        ...
+
+    @abstractmethod
+    def follow_link(self, **kwargs) -> None:
+       ...
+
+    def goto(self, handle: str) -> None:
+        self._cursor = handle
+
+
+class HandleOnlyTraverseEngine(TraverseEngine):
+    def get(self) -> str:
+        return self._cursor
+
+    def get_links(self, **kwargs) -> ListIterator:
+        incoming_links = self._get_incoming_links()
+
+        if not incoming_links:
+            return ListIterator([])
+
+        filtered_links = self._filter_links(
+            links=incoming_links,
+            link_type=kwargs.get('link_type'),
+            cursor_position=kwargs.get('cursor_position'),
+            target_type=kwargs.get('target_type'),
+            handles_only=True,
+        )
+
+        return ListIterator(filtered_links)
+
+    def get_neighbors(self, **kwargs) -> List[str]:
+        filtered_links_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
+            target_type=kwargs.get('target_type')
+        )
+
+        targets_set = {target for link in filtered_links_iterator for target in link['targets']}
+
+        if not targets_set:
+            return []
+
+        targets_set.discard(self._cursor)
+
+        return list(targets_set)
+
+    def follow_link(self, **kwargs) -> None:
+        filtered_links_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
+            target_type=kwargs.get('target_type')
+        )
+
+        filtered_links = [link for link in filtered_links_iterator]
+
+        if not filtered_links:
+            return
+
+        unique_path = kwargs.get('unique_path', False)
+
+        if unique_path and len(filtered_links) > 1:
+            raise MultiplePathsError(
+                message='Unable to follow the link. More than one path found',
+                details=f'{len(filtered_links)} paths',
+            )
+
+        link = choice(filtered_links)
+
+        if self._cursor in link['targets']:
+            link['targets'].remove(self._cursor)
+
+        handle = choice(link['targets'])
+
+        self.goto(handle)
```

Comment:
`goto()` should not be used here. As soon as we have a cache in place, `goto()` will be a much more expensive operation than simply updating current cursor. So I suggest to just update the cursor here right away.